File: src/test/java/net/bbelovic/adventofcode/day3/Day3PuzzlePart1Test.java
package net.bbelovic.adventofcode.day3;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

class Day3PuzzlePart1Test {

    @ParameterizedTest
    @MethodSource("testData")
    void
    should_return_number_of_locations_visited_at_least_once(String input, int expectedLocationCount) {
        Day3PuzzlePart1 day3Part1 = new Day3PuzzlePart1();
        var actualLocationCount = day3Part1.solve(input);
        assertEquals(expectedLocationCount, actualLocationCount);
    }

    static Iterable<Arguments> testData() {
        return List.of(
                Arguments.of("^v", 2, 3),
                Arguments.of("^>v<", 4, 3),
                Arguments.of("^v^v^v^v^v", 2, 11)
        );
    }
}

File: aoc-2018/src/test/kotlin/net/bbelovic/adventofcode/year2018/day3/Day3PuzzlePart1Test.kt
package net.bbelovic.adventofcode.year2018.day3

import net.bbelovic.adventofcode.utils.dsl.testDataDsl
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource

class Day3PuzzlePart1Test {
    @ParameterizedTest
    @MethodSource("testData")
    fun `should compute fabric overlaps`(rectangles: List<String>, expectedOverlap: Long) {
        val puzzle = Day3PuzzlePart1()
        val actualOverlap = puzzle.solve(rectangles)
        assertEquals(expectedOverlap, actualOverlap)
    }

    @Test
    fun `should throw exception for invalid input`() {
        assertThrows(IllegalArgumentException().javaClass) { RectangleParser.parse("@ 335,901: 28x17") }
        assertThrows(IllegalArgumentException().javaClass) { RectangleParser.parse("# @ 335,901: 28x17") }
        assertThrows(IllegalArgumentException().javaClass) { RectangleParser.parse("#1 @ ,901: 28x17") }
        assertThrows(IllegalArgumentException().javaClass) { RectangleParser.parse("#1 @ 335,901: ") }
    }

    companion object {
        @JvmStatic
        private fun testData() = testDataDsl {
            testArguments (listOf("#1 @ 0,0: 1x3", "#2 @ 0,0: 3x1"), 1L)
            testArguments (listOf("#1 @ 0,0: 3x1", "#2 @ 0,2: 3x1"), 0L)
            testArguments (listOf("#1 @ 0,0: 2x2", "#2 @ 1,1: 2x2"), 1L)
            testArguments (listOf("#1 @ 0,0: 1x1", "#2 @ 0,0: 1x1", "#3 @ 0,0: 1x1"), 1L)
        }
    }
}